response_syncrepl.go

```diff
@@ -0,0 +1,156 @@
+package ldap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	ber "github.com/go-asn1-ber/asn1-ber"
+)
+
+type syncreplResponse struct {
+	sr *searchResponse
+}
+
+// Entry returns an entry from the given search request
+func (r *syncreplResponse) Entry() *Entry {
+	return r.sr.entry
+}
+
+// Referral returns a referral from the given search request
+func (r *syncreplResponse) Referral() string {
+	return r.sr.referral
+}
+
+// Controls returns controls from the given search request
+func (r *syncreplResponse) Controls() []Control {
+	return r.sr.controls
+}
+
+// Err returns an error when the given search request was failed
+func (r *syncreplResponse) Err() error {
+	return r.sr.err
+}
+
+// Next returns whether next data exist or not
+func (r *syncreplResponse) Next() bool {
+	return r.sr.Next()
+}
+
+func (r *syncreplResponse) start(ctx context.Context, searchRequest *SearchRequest) {
```

Comment:
> I suggest to remove this duplicate and use a control with SearchAsync instead.

I also tried to implement the syncrepl feature into the searchResponse struct with SearchAsync() method. And then, I found we have to change several code places. Because existent search response packets and syncrepl search response packets are a little different. Not only ApplicationIntermediateResponse.

For example, compare handling ApplicationSearchResultEntry of the response packet. The searchResponse doesn't handle controls.

* searchResponse

```go
				switch packet.Children[1].Tag {
				case ApplicationSearchResultEntry:
					r.ch <- &SearchSingleResult{
						Entry: &Entry{
							DN:         packet.Children[1].Children[0].Value.(string),
							Attributes: unpackAttributes(packet.Children[1].Children[1].Children),
						},
					}
```

But, the syncreplResponse expects an entry with a control. To implement the syncrepl functions, I thought it might change existing behavior (I'm not sure whether the existing behavior is correct or not). Do you think this is acceptable?

* syncreplResponse

```go
				case ApplicationSearchResultEntry:
					result := &SearchSingleResult{
						Entry: &Entry{
							DN:         packet.Children[1].Children[0].Value.(string),
							Attributes: unpackAttributes(packet.Children[1].Children[1].Children),
						},
					}
					if len(packet.Children) != 3 {
						r.sr.ch <- result
						continue
					}
					controlPacket := packet.Children[2].Children[0]
					decoded, err := DecodeSyncreplControl(controlPacket)
					if err != nil {
						werr := fmt.Errorf("failed to decode search result entry: %w", err)
						result.Error = werr
						r.sr.ch <- result
						return
					}
					result.Controls = append(result.Controls, decoded)
					r.sr.ch <- result
```

I think adding the syncrepl feature makes increasing some check conditions to work them together (both normal search and syncrepl consumer). I can implement them, but I'm not sure whether some side effects affect a normal search or not. This is a trade-off between maintenance cost and compatibility.

Reply:
I agree that reducing unnecessary code duplication is a worthy goal, in this case I'm okay with a distinct synreplResponse

Reply:
@cpuschma I integrated the source code for maintainability. I did change the existing process carefully, so it is almost an additional function. There are only three parts to be changed. That is few than I expected. So I think you are right about maintainability.

Could you review it again?